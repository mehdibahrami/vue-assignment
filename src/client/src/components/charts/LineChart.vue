<template>
  <!-- Apex Line Chart -->
  <apexchart
    :type="type"
    :options="this.options"
    ref="chart"
    :series="this.series"
  >
  </apexchart>
</template>

<script>
import VueApexCharts from "vue3-apexcharts";

export default {
  name: "LineChart",
  components: {
    apexchart: VueApexCharts,
  },
  props: {
    title: {
      type: String,
    },
    data: {
      type: Object,
      require: true,
    },
    options: {
      type: Object,
      required: true,
    },
    interval: {
      type: Number,
    },
    type: {
      type: String,
      default: "line",
    },
    setColor: {
      type: Function,
    },
  },
  data() {
    return {
      series: [{ name: this.title, data: [] }],
    };
  },
  mounted() {
    // If it is a realtime chart(interval props has value) update the data
    if (this.interval && this.interval > 0) {
      // Updates the data according to the interval time period
      this.setData();

      // Reset data every 5 minutes to prevent memory leak (https://github.com/apexcharts/apexcharts.js/issues/605)
      this.resetData();
    }
  },
  methods: {
    setData: function () {
      setInterval(() => {
        // If the x and y values are not null then check other parameters for update chart series
        if (this.data.x && this.data.y) {
          const dataLength = this.series[0].data.length;

          // If series contains at least one element then check new time value(this.data.x) should be greater than last time in the array
          // If it is not we should reset the chart
          if (dataLength > 1) {
            const lastData = this.series[0].data[dataLength - 1];
            if (lastData[0] > this.data.x) {
              this.series[0].data = [];
            }
          }

          // Push the new data to series array and update chart
          this.series[0].data.push([this.data.x, this.data.y]);
          this.updateChart();

          // Set chart color based on current value
          if (this.setColor) this.setColor();
        }
      }, this.interval);
    },
    updateChart: function () {
      // Update chart series
      this.$refs.chart.updateSeries([
        {
          data: this.series[0].data,
        },
      ]);
    },
    resetData: function () {
      setInterval(() => {
        const dataLength = this.series[0].data.length;

        // If series has more than 100 elements delete 70 of the oldest
        if (dataLength > 100) {
          this.series[0].data = this.series[0].data.slice(
            dataLength - 70,
            dataLength
          );

          this.$refs.chart.updateSeries([
            {
              data: [this.series[0].data],
            },
          ]);
        }
      }, 300000);
    },
  },
};
</script>
