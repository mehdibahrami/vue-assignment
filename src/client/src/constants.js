// Maximum charge
export const MAX_CHARGE = 100;
// Maximum amount of middle charge level
export const MID_CHARGE = 65;
// Maximum amount of low charge level
export const LOW_CHARGE = 30;
// Minimum charge level
export const MIN_CHARGE = 0;
// Maximum speed
export const MAX_SPEED = 100;
// SoC line chart interval
export const SOC_LINE_CHART_INTERVAL = 100;
// Minimum speed
export const MIN_SPEED = 0;
// Maximum allowed speed
export const SPEED_LIMIT = 50;
// Speed line chart interval
export const SPEED_LINE_CHART_INTERVAL = 100;
