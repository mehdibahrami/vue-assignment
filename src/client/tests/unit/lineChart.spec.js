import { mount } from "@vue/test-utils";
import LineChart from "@/components/charts/LineChart.vue";
import {
  speedLineChartOptions,
  setSpeedLineChartColor,
} from "@/components/charts/chartsOptions.js";
import { SPEED_LINE_CHART_INTERVAL } from "@/constants";

describe("LineChart", () => {
  let wrapper;

  it("renders correctly", () => {
    // shallow mounting the main component
    wrapper = mount(LineChart, {
      props: {
        title: "Speed",
        data: { x: 1511436339000, y: 4 },
        options: speedLineChartOptions,
        interval: SPEED_LINE_CHART_INTERVAL,
        setColor: setSpeedLineChartColor,
      },
    });

    expect(wrapper.findComponent({ name: "apexchart" }).exists()).toBeTruthy();
    expect(wrapper.find(".vue-apexcharts").exists()).toBeTruthy();
    expect(wrapper.element).toMatchSnapshot();
  });
});
