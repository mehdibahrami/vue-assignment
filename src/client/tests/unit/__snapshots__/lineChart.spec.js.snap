// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart renders correctly 1`] = `
<div
  data-v-app=""
>
  
  <!-- Apex Line Chart -->
  <div
    class="vue-apexcharts"
  />
  
</div>
`;
